Open Problem: Enhanced Bitswap/GraphSync with more Network Smarts (#9)
ipfs/research: OPEN_PROBLEMS/ENHANCED_BITSWAP_GRAPHSYNC.md

# Enhanced Bitswap/GraphSync with more Network Smarts

## Short Description
> In one sentence or paragraph.

> Once IPFS has resolved the CID to a peerID and further to the exact location of the node that  stores the requested content, it then uses a very simple protocol called bitswap to exchange content between the requestor and the server.

Reviewer: This is broadly correct, however Bitswap also participates in the discovery phase:
- IPFS asks Bitswap for the blocks corresponding to a set of CIDs
- Bitswap sends a request for the CIDs to all connected peers
- If no connected peers have a particular block, Bitswap searches for more peers by querying the DHT for the first CID (usually the root node in a DAG)

> Although bitswap is simple and generally works, *its performance is significantly suboptimal*. This is mainly due to the fact that a node cannot request a subgraph of the DAG and results in many round-trips in order to “walk down” the DAG. The current operation of bitswap is also very often linked to duplicate transmission and receipt of content which overloads both the end nodes but also the network.

Reviewer 2: ```suggestion
Although bitswap is simple and generally works, *its performance is suboptimal*. This is mainly due to the fact that a node cannot request a subgraph of the DAG and results in many round-trips in order to “walk down” the DAG. The current operation of bitswap is also very often linked to duplicate transmission and receipt of content which overloads both the end nodes and the network.
```

Language fixes.

## Long Description

> In order to synchronise a DAG between untrusted nodes, bitswap is exploiting the content-addressability feature of IPFS. This means that bitswap starts with the root hash of the DAG and once it has the data, it rehashes the data to verify that it results to the same hash. The node can now trust this block and hence, can continue with the rest of the blocks in the DAG (aka walk the DAG), until it gets the whole DAG and the data associated with it.

Reviewer 2: ```suggestion
In order to synchronise a DAG between untrusted nodes, bitswap is exploiting the content-addressability feature of IPFS. This means that bitswap starts with the root hash of the DAG. Once it fetches the data, it verifies that its hash matches. The node can now trust this block and can thus continue with the rest of the blocks in the DAG (aka walk the DAG) until it gets the complete DAG and the data associated with it.
```
Language + clarity.

> In the current implementation of bitswap, requesting nodes send their WANT lists to all the peers they are directly connected to. This inevitably results in potentially duplicate traffic travelling back to the receiving node. When the receiving node has received the block(s) it asked for, it sends a CANCEL message to its peers to let them know that the block(s) is not in its WANT list anymore.

Reviewer 2: ```suggestion
In the current implementation of bitswap, requesting nodes send their WANT lists to all the peers they are directly connected to. This results in potentially duplicate data being sent back to the receiving node. When the receiving node has received the block(s) it asked for, it sends a CANCEL message to its peers to let them know that the block(s) is not in its WANT list anymore.
```
Language

Reviewer 2: To be clear, is that "(s)" correct? Do multiple cancellations get collated into the same CANCEL message? The go-bitswap readme (https://github.com/ipfs/go-bitswap#requesting-blocks) says that is indeed the case, both for requests and cancellations. However, the asynchronous nature of block retrieval (and, consequently, cancellation) would make this unlikely unless outgoing messages are queued for a _long_ (for some definition of long) time. There is an implied efficiency trade-off here, between reducing outgoing signalling traffic and reducing redundant inbound traffic.

> If none of the directly connected peers have any of the WANT list blocks, bitswap falls back to the DHT to find the requested content. Obviously, this results in long delays to get to the actual peer that stores the requested content.

Reviewer 2: ```suggestion
If none of the directly connected peers have any of the WANT list blocks, bitswap falls back to the DHT to find the requested content. This results in long delays to get to a peer that stores the requested content.
```
Language + clarity

> Furthermore, once the recipient node starts receiving content from multiple peer nodes, it prunes down the long-latency peers and keeps the one to which the RTT is the shortest. Instead, current proposals within the IPFS ecosystem are considering optimising for throughput. Although this has its advantages, which one is the ideal to follow is not clear at this point.

Reviewer 2: Which one of which? Are we contrasting optimising for throughput vs. efficiency? Time to first block? Because, other than the pruning, the current system seems very much optimised for throughput over efficiency too.

Reviewer 2: ```suggestion
Once the recipient node starts receiving content from multiple peer nodes, it prunes down the long-latency peers and keeps the one to which the RTT is the shortest. Current proposals within the IPFS ecosystem are considering keeping the node with the highest throughput instead. It is not clear at this point which is the best approach.
```

Ok, this became clearer when reading the related work below. See if it makes sense.

It is important to highlight that bitswap is a message-oriented protocol and _not_ a request-response protocol. That is, it sends messages including WANT lists and  is then waiting for the blocks in the WANT list to be delivered back, as and when the discovery protocol manages to find the blocks in the WANT list.

This process of bitswap results in the following major problems:

- **Many Roundtrips:** The most pressing issue is that requesting nodes have to make several roundtrips to walk the DAG, especially given that IPFS is a network run by untrusted nodes.

> - **Duplicate Data:** Bitswap is operating at the block level. In turn, this means that every block has to be requested from each peer that the node is connected to. This process results in duplicate traffic towards the data requestor, unnecessary load for the peers, and high overhead for the network.

Reviewer: In the current release of Bitswap there is the concept of Sessions, used to keep track of peers that have some of the blocks that the Session is interested in:
- When IPFS asks Bitswap for the blocks corresponding to a set of CIDs, Bitswap
  - creates a new session
  - starts a discovery phase to find the peers with a subset of the blocks
- As peers are discovered they are remembered by the Session
- The Session sends subsequent requests only to those peers
- If there is a timeout, the Session tries to discover more peers through the DHT

> - **Duplicate Data:** as mentioned, bitswap is operating at the block level. In turn, this means that every block has to be requested from each peer that the node is connected to. This process results in duplicate traffic towards the data requestor unnecessary load for the peers and high overhead for the network.

Reviewer 2: ```suggestion
- **Duplicate Data:** Bitswap is operating at the block level. In turn, this means that every block has to be requested from each peer that the node is connected to. This process results in duplicate traffic towards the data requestor, unnecessary load for the peers, and high overhead for the network.
```
Language.

Reviewer 2: Outstanding issue: It does not follow from operating at the block level that there has to be traffic duplication or that every block has to be requested from every peer. At most, it follows that the request volume is larger because the request unit is smaller. The retrieval strategy (and consequent duplication ratio) seems to be a separate issue.

## State of the Art

> This survey on the State of the Art is not by any means complete, however, it should provide a good entry point to learn what are the existing work. If you have something that is fundamentally missing, please consider submitting a PR to augment this survey. 

### Within the IPFS Ecosystem
> Existing attempts and strategies

The IPFS community has long been aware of the issues of bitswap and has experimented (at least conceptually) with potential solutions. We are discussing some of them here:

- *Parallelise the DAG walk.* Inherently, DAGs have one root node and then split into binary trees. Doing one walk for each of the branches results in the same number of round trips, but reduces the actual delay in half. Clearly, this approach does not apply for single-branch graphs, such as blockchains.

> - *DAG Block Interconnection.* Although bitswap does not/cannot recognise any relationship between different blocks of the same DAG, a requesting node can artificially continuously ask for subsequent blocks of the same block the node that it found the previous block. This approach intuitively assumes that a node that had a block of a DAG is very likely to have subsequent ones. This is often referred to as “session” between the peers that have provided some part of the DAG.

Reviewer 2: ```suggestion
- *DAG Block Interconnection.* Although bitswap does not/cannot recognise any relationship between different blocks of the same DAG, a requesting node can ask a node that provided a previous block for subsequent blocks of the same DAG. This approach intuitively assumes that a node that has a block of a DAG is very likely to have others. This is often referred to as “session” between the peers that have provided some part of the DAG.
```
I assume this is what was meant.

Reviewer 2: My suggestion keeps the wording "[ask] for subsequent blocks of the same DAG". What was written before was "ask for subsequent blocks of the same block the node that it found the previous block", which doesn't really parse -- so it may or may not be equivalent.

- *Latency vs Throughput.* Bitswap is currently sorting peers by latency, i.e., it is pruning down the connections that incur higher latency. It has been suggested that this is changed to maximise throughput (i.e., keep the pipe full).

- *Coding and Reed Solomon Codes.* This is a very efficient way to reduce storage space in general (trusted or untrusted) caching systems. The area of coded caching has seen significant development lately in the research community, hence, it is discussed further down.

> - *GraphSync through IPLD selectors:* GraphSync is a specification of a protocol to synchronise graphs across peers. Graphsync answers the question of “how do we succinctly identify selectors connected to the root”. In doing that, it develops WANT lists of selectors instead of blocks and requests have selectors instead of hashes. Furthermore, in contrast to bitswap which is message-oriented, graph sync includes responses and hence, it is closer to a request-response protocol. If designed carefully, Graphsync can be very effective, but there are challenges that are still unresolved: i) it is more difficult to parallelise (see above), given that GraphSync operates in terms of requests and not blocks, ii) it can potentially be much easier to carry DDoS attacks with queries than with requests for blocks.

Reviewer 2: ```suggestion
- *GraphSync through IPLD selectors:* GraphSync is a specification of a protocol to synchronise graphs across peers. Graphsync answers the question of “how do we succinctly identify selectors connected to the root”. In doing that, it develops WANT lists of selectors instead of blocks. Furthermore, in contrast to bitswap, which is message-oriented, GraphSync includes responses and is thus closer to a request-response protocol. If designed carefully, GraphSync can be very effective but there are still unresolved challenges: i) it is more difficult to parallelise (see above), given that GraphSync operates in terms of requests and not blocks, ii) it can potentially be much easier to carry DDoS attacks with queries than with requests for blocks.
```
Language

> - *WANT Potential and HAVE message.* There have been several optimisation proposals, such as the implementation of a HAVE message before actually sending the requested block back. Although this immediately reduces duplicate traffic, it introduces one extra RTT, which might not be desirable in some cases, but might not have any negative impact in many others.

Reviewer: Just fleshing this out a little:
- The new message types are
  - HAVE: a node can indicate that it has a block (helps with reducing duplication, particularly during the discovery phase)
  - DONT_HAVE: a node can immediately indicate that it does not have a block (in the current implementation the requesting node has to wait for a timeout)

Note that the proof-of-concept we are building sends a combination of
- want-block (give me a block if you have it)
  This is to try to avoid the RTT
- want-have (let me know if you have the block)
  This is to avoid duplication and help disseminate knowledge of block distribution in the network

The want-block messages are directed to the peers that are most likely to have them (sent a HAVE message for current CID or sent block / HAVE for previous requests)

- Golang Implementation of Bitswap: https://github.com/ipfs/go-bitswap
- GraphSync: https://github.com/ipld/specs/blob/master/block-layer/graphsync/graphsync.md
- Bitswap Latency Measurement: https://github.com/ipfs/go-bitswap/issues/166
- Bitswap Sessions Exploration: https://github.com/ipfs/go-bitswap/issues/165
- Bitswap Protocol Extensions: https://github.com/ipfs/go-bitswap/issues/186
- Reed-Solomon layer over IPFS: https://github.com/ipfs/notes/issues/196
- Bitswap Request Sharding: https://github.com/ipfs/go-bitswap/issues/167

> 

Reviewer: Bitswap POC with HAVE / DONT_HAVE: https://github.com/ipfs/go-bitswap/pull/189

The approaches that bittorrent has used need to be evaluated, but always keeping in mind that its architecture is different and based on a (logically) centralised tracker.

Coded caching on the other hand incurs some overhead, which might not be ideal for several use-cases of the IPFS system (e.g., popular content), but very convenient for others (cold storage).

## Solving this Open Problem

### What is the impact

Bitswap is a central component of the IPFS system and a protocol that generally influences the performance of IPFS as a whole. Designing and integrating smart algorithmic solutions to improve its performance can be instrumental for the performance and adoption of IPFS. It is worth highlighting that bitswap allows ample space for improvement, hence, several different solutions are encouraged and will be considered. Addressing sub-graphs of the DAG through IPLD as GraphSync is attempting to do in a secure manner is a first necessary step.

### What defines a complete solution?
> What hard constraints should it obey? Are there additional soft constraints that a solution would ideally obey?

> First and foremost, any complete solution should account for extensibility as the IPFS system needs to scale up and more applications are implemented on top. The active number of users of IPFS is increasing exponentially and the requests submitted to the network are following accordingly. That said, a complete solution should account for those numbers.

Reviewer 2: I'm not sure unbounded exponential scaling is a realistic goal. Would be good to put some order of magnitude here, especially given the reference to "those numbers".

Reviewer: I would add that ideally IPFS should dynamically adapt to different environments, analogously to how TCP works in a data center and also works on the broader internet

# Enhanced Bitswap/GraphSync with more Network Smarts

## Short Description

> > In one sentence or paragraph.

Author: @yiannisbot this short description has 3 paragraphs, that's what the Long description is for :)

## Long Description

In order to synchronise a DAG between untrusted nodes, bitswap is exploiting the content-addressability feature of IPFS. This means that bitswap starts with the root hash of the DAG. Once it fetches the data, it verifies that its hash matches. The node can now trust this block and can thus continue with the rest of the blocks in the DAG (aka walk the DAG) until it gets the complete DAG and the data associated with it.

In the current implementation of bitswap, requesting nodes send their WANT lists to all the peers they are directly connected to. This results in potentially duplicate data being sent back to the receiving node. When the receiving node has received the block(s) it asked for, it sends a CANCEL message to its peers to let them know that the block(s) is not in its WANT list anymore.

If none of the directly connected peers have any of the WANT list blocks, bitswap falls back to the DHT to find the requested content. This results in long delays to get to a peer that stores the requested content.

> Once the recipient node starts receiving content from multiple peer nodes, it prunes down the long-latency peers and keeps the one to which the RTT is the shortest. Current proposals within the IPFS ecosystem are considering keeping the node with the highest throughput instead. It is not clear at this point which is the best approach.

Reviewer 3: Not exactly.

* We currently prune to peers that _have_ the content, then prioritize sending wants to peers with lower latencies. We still send wants to all peers (IIRC).
* The plan is to change that second part to: prioritize sending wants to peers with the least amount of queued work.

Author: @yiannisbot did you take @Stebalien's review in this comment?

This process of bitswap results in the following major problems:

- **Many Roundtrips:** The most pressing issue is that requesting nodes have to make several roundtrips to walk the DAG, especially given that IPFS is a network run by untrusted nodes.
- **Duplicate Data:** Bitswap is operating at the block level. In turn, this means that every block has to be requested from each peer that the node is connected to. This process results in duplicate traffic towards the data requestor, unnecessary load for the peers, and high overhead for the network.

In the latest release of Bitswap there is an optimisation of the above procedure that incorporates the concept of Sessions. Sessions are used to keep track of peers that have some of the blocks that the Session is interested in. In particular:

- When IPFS asks Bitswap for the blocks corresponding to a set of CIDs, Bitswap:
  - creates a new session
  - starts a discovery phase to find the peers with a subset of the blocks.
- As peers are discovered they are remembered by the Session. Subsequent requests are sent only to those peers.

> - If there is a timeout, the Session tries to discover more peers through the DHT.

Reviewer 3: Third point: bitswap _divides_ the request between peers in a session. That is, it doesn't ask every peer for every block, it asks a subset of the peers in the session (usually 1-2) for each block.

### Within the IPFS Ecosystem
> Existing attempts and strategies

The IPFS community has long been aware of the issues of bitswap and has experimented (at least conceptually) with potential solutions. We are discussing some of them here:

- *Parallelise the DAG walk.* Inherently, DAGs have one root node and then split into binary trees. Doing one walk for each of the branches results in the same number of round trips, but reduces the actual delay in half. Clearly, this approach does not apply for single-branch graphs, such as blockchains.

- *DAG Block Interconnection.* Although bitswap does not/cannot recognise any relationship between different blocks of the same DAG, a requesting node can ask a node that provided a previous block for subsequent blocks of the same DAG. This approach intuitively assumes that a node that has a block of a DAG is very likely to have others. This is often referred to as “session” between the peers that have provided some part of the DAG.

> - *Latency vs Throughput.* Bitswap is currently sorting peers by latency, i.e., it is pruning down the connections that incur higher latency. It has been suggested that this is changed to maximise throughput (i.e., keep the pipe full).

Reviewer 3: It's not really either/or. Really, we should:

* Optimize for _latency_ when traversing deep/narrow DAGs (e.g., a blockchain/path). Lower latency means we learn about the next node faster.
* Optimize for throughput when traversing a wide DAG in parallel.

- Do incentives build robustness in BitTorrent?, NSDI 2007, https://www.usenix.org/legacy/events/nsdi07/tech/piatek/piatek.pdf
- Modeling and performance analysis of BitTorrent-like peer-to-peer networks, ACM CCR 2004 http://conferences.sigcomm.org/sigcomm/2004/papers/p444-qiu1.pdf
- Analyzing and Improving a BitTorrent Network’s Performance Mechanisms, Infocom 2006, http://jmvidal.cse.sc.edu/library/bharambe06a.pdf

*Coded Caching*

There have been significant research efforts lately in the area of coded caching. The main concept has been proposed in 1960s in the form of error correction and targeted the area of content delivery over wireless, lossy channels. It has been known as Reed-Solomon error correction. Lately, with seminal works such as “Fundamental Limits of Caching”, Niesen et. al. have proposed the use of coding to improve caching performance. In a summary, the technique works as follows: if we have a file that consists of 10 chunks and we store all 10 chunks in the same or different memories/nodes, then we need to retrieve those exact 10 chunks in order to reconstruct the file.

> In contrast, according to the coded caching theory, before storing the 10 chunks we encode the file using erasure codes. This results in some number of chunks x>10, say 13, for the sake of illustration. This clearly results in more data produced after adding codes to the original data. However, when attempting to retrieve the original file, a user needs to collect *any 10 of those 13 chunks*. By doing so, the user will be able to reconstruct the original file, without needing to get all 13 chunks. Although such approach does not save bandwidth (we still need to reconstruct 10 chunks of equal size to the original one), it makes the network more resilient to nodes being unavailable. In other words, in order to reconstruct the original file without coding, all 10 of the original peers that store a file have to be online and ready to deliver the chunks, whereas in the coded caching case, any 10 out of the 13 peers need to be available and ready to provide the chunks. Blind replication of the original chunks will not provide the same benefit, as the number of peers will need to be much higher (at least 20 as compared to 13) in order to operate with the same satisfaction ratio.

Reviewer 3: IIRC, if I erasure encode my file, I can't reconstruct a _part_ of the file without having the minimum number of chunks. Is this correct?

If so, I'm not sure if this buys us anything. Peers will likely either have or not have all the chunks necessary to reconstruct a file; having one chunk will be highly correlated with having the rest.

TL;DR: chunk deletion _is not_ random. Yes, disks can fail but that should be handled at a lower layer.

- *Parallelise the DAG walk.* Inherently, DAGs have one root node and then split into binary trees. Doing one walk for each of the branches results in the same number of round trips, but reduces the actual delay in half. Clearly, this approach does not apply for single-branch graphs, such as blockchains.

- *DAG Block Interconnection.* Although bitswap does not/cannot recognise any relationship between different blocks of the same DAG, a requesting node can ask a node that provided a previous block for subsequent blocks of the same DAG. This approach intuitively assumes that a node that has a block of a DAG is very likely to have others. This is often referred to as “session” between the peers that have provided some part of the DAG.

- *Latency & Throughput Optimisation.* Bitswap is currently sorting peers by latency, i.e., it is prioritising the connections that incur lower latency. Ideally, a hybrid approach needs to be implemented, where prioritisation based on latency applies to narrow but deep DAGs (e.g., blockchains), whereas prioritisation based on higher throughput applies to wide DAGs that are being traversed in parallel.

> - *Coding and Reed Solomon Codes.* This is a very efficient way to reduce storage space in general (trusted or untrusted) caching systems. The area of coded caching has seen significant development lately in the research community, hence, it is discussed further down.

Author: I'm not familiar with this experiment in the IPFS Ecosystem, where did it happen? Can you add links to it?

Author: Oh I see. https://github.com/ipfs/notes/issues/196 doesn't mean we tried it :).

## Other

### Existing Conversations/Threads

- [Batch gets](https://github.com/ipfs/notes/issues/285)
- [Move giant files/datasets at Level1 + Level2 ](https://github.com/ipfs/notes/issues/218)
- [Reed-Solomon layer over IPFS](https://github.com/ipfs/notes/issues/196)
- [Tracking progress and other data](https://github.com/ipfs/notes/issues/107)
- [Advanced Pinning](https://github.com/ipfs/notes/issues/49)
- [bandwidth estimators](https://github.com/ipfs/notes/issues/30)
- [Alternative BitSwap strategies](https://github.com/ipfs/notes/issues/20)
- [2-Hop Bitswap](https://github.com/ipfs/notes/issues/386)
- [IPFS Spy](https://github.com/jimpick/go-ipfs/blob/jim/network-logging/README.ipfs-spy.md)
- [Bitswap Simulator](https://github.com/heems/bssim)
- [bsdash - dashboard for go-ipfs bitswap](https://www.npmjs.com/package/bsdash)

- Presentation of bitswap in Sept 2019 IPFS Weekly Call: https://www.youtube.com/watch?v=G_Q7iTpwYQU&list=PLuhRWgmPaHtSGRSHdU9dbsukHKlihZZAe&index=4

### Extra notes

> [qri.io](https://qri.io/): a data transfer mechanism using IPFS components to keep track of blocks in a DAG using Manifest files (similar to bittorrent magnet files) - https://github.com/qri-io/dag

Author: This should go as "one of the experiments within the IPFS Ecosystem", it is a tool that uses IPFS and its APIs for faster syncs.

### Within the IPFS Ecosystem
> Existing attempts and strategies

### Within the broad Research Ecosystem
> How do people try to solve this problem?

### Known shortcommins of existing solutions
> What are the limitations on those solutions?

## Solving this Open Problem

### What is the impact

### What defines a complete solution?
> What hard constraints should it obey? Are there additional soft constraints that a solution would ideally obey?

## Other

### Existing Conversations/Threads

- [Batch gets](https://github.com/ipfs/notes/issues/285)
- [Move giant files/datasets at Level1 + Level2 ](https://github.com/ipfs/notes/issues/218)
- [Reed-Solomon layer over IPFS](https://github.com/ipfs/notes/issues/196)
- [Tracking progress and other data](https://github.com/ipfs/notes/issues/107)
- [Advanced Pinning](https://github.com/ipfs/notes/issues/49)
- [bandwidth estimators](https://github.com/ipfs/notes/issues/30)
- [Alternative BitSwap strategies](https://github.com/ipfs/notes/issues/20)
- [Bitswap Simulator](https://github.com/heems/bssim)
- [bsdash - dashboard for go-ipfs bitswap](https://www.npmjs.com/package/bsdash)

> 

Author: @jimpick where can I find your ipfs-spy project to add here?

Reviewer 4: https://github.com/jimpick/go-ipfs/blob/jim/network-logging/README.ipfs-spy.md

It’s a very early stage experiment currently...
